### PR TITLE
New translator: HCSP

### PR DESCRIPTION
**Pull request by jomuller**
As suggested in the [zotero dev Google group](https://groups.google.com/forum/#!topic/zotero-dev/WXIVULPDYVc), I send a pull request for a new translator.

This translator is for the website ["Haut Conseil de la Santé Publique"](http://www.hcsp.fr/explore.cgi/avisrapports) widely used in France for healthcare. It's the first translator I wrote, then there is most likely some defect in this script. I tested it on a lot of pages (like this one) in this website, and it worked well. Please, feel free to criticize if there is something bad.

If it reaches the quality standards, should it be possible to make available this translator for the others Zotero users?

### CODE REVIEW

File: HCSP.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
+{
+    "translatorID": "3e375db7-5f6e-4861-ab8b-3ddb0f4accc9",
+    "label": "HCSP",
+    "creator": "Joris Muller",
+    "target": "http://www.hcsp.fr",
```
Comment (adam3smith) [inline HCSP.js]: ```^https?//www\.hcsp\.fr/```
Make more efficient, allow for a current or future https version, escape ```.``` and end target with a ```/``` to work with proxies.

File: HCSP.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+// Usage of Zotero.Utilities.xpath() function. Exist only since Zotero 2.1
+// https://www.zotero.org/support/dev/client_coding/changes_in_zotero_2.1#xpath_utility_functions
+// Then minimal version of Zotero needed is 2.1
+
+// TODO: Use camelStyle for variables to be consistent with Zotero coding style
+
+function detectWeb(doc, url) {
+    /* The URLs have upper or lower case on the website
+    depending the way you access to a page */
+    if (url.match("[Ee]xplore.cgi/avisrapportsdomaine")) {
```
Comment (adam3smith) [inline HCSP.js]: if you're matching on a literal string, just use
```.indexOf("myString") != -1``` which is much more efficient -- particularly important for the detect function which runs a lot.

File: HCSP.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        return 'report';
+    } else if (url.match("[Ee]xplore.cgi/avisrapports")) {
```
Comment (adam3smith) [inline HCSP.js]: ditto

File: HCSP.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        // NOTE: "for of loop" are quite new and works only with newest browsers
+        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
+        var allItems = {};
+        for (var nextTitle of titles) {
+            allItems[nextTitle.href] = nextTitle.textContent;
+        }
+        
+        // Let the user choose the items he wants
+        var itemsSelected;
+        itemsSelected = Zotero.selectItems(allItems);
```
Comment (adam3smith) [inline HCSP.js]: ```Zotero.selectItems``` should be run as a callback function. See e.g. https://github.com/zotero/translators/blob/d82c1efd2ca75af0caa0d94884e38d2e6afbc210/Dryad%20Digital%20Repository.js#L60 
for an example that otherwise does the same you do.

File: HCSP.js [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            articles.push(one_item)
+        }
+
+    } else {
+        // Page of the article : get the URL
+        articles = [url]
+    }
+
+    Zotero.Utilities.processDocuments(articles, scrape, function(){Zotero.done();});
+    Zotero.wait();
```
Comment (adam3smith) [inline HCSP.js]: also see the translator above for how to simplify this: ```Zotero.done``` and ```Zotero.wait``` are no longer needed.

File: HCSP.js [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+    var namespace = doc.documentElement.namespaceURI;
+    // Resolver for XPaths
+
+    // init a new Zotero's item
+    var newItem = new Zotero.Item("report");
+
+    /*---------- Constants items ----------*/
+    newItem.publicationTitle = "Rapport de l'HCSP";
+    newItem.language = "fr";
```
Comment (adam3smith) [inline HCSP.js]: ```fr-FR``` (presumably)

File: HCSP.js [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        // Titles and URLs of each item are on the third column of the only table
+        // This column is the only one with links (<a>)
+        var xpath_titles = '//tbody/tr/td/a';
+        var titles = ZU.xpath(doc, xpath_titles);
+        
+        // Put all the titles and their URLs in the items object
+        // NOTE: "for of loop" are quite new and works only with newest browsers
+        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
+        var allItems = {};
+        for (var nextTitle of titles) {
```
Comment (adam3smith) [inline HCSP.js]: I'm not sure, but given that ```for...of``` doesn't run in IE, I think we might prefer a different loop here. We usually just do
```for(var i = 0; i<titles.length;i++)``` which has the added benefit of being super efficient.

File: HCSP.js [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+    for (var nextResult of xresultTags) {
+        var actualTag = nextResult.textContent;
+        var actualTagTrimmed = actualTag.trim();
+        newItem.tags.push(actualTagTrimmed);
+    }
+
+    /* --------- abstract -----------*/
+    var xpath_abstract = '//div[@class="avistexte"]';
+    var abstract_note = ZU.trimInternal(ZU.xpathText(doc, xpath_abstract, namespace, "\n"));
```
Comment (adam3smith) [inline HCSP.js]: why do you need the namespace here?
Also, you should test for the existence of ```ZU.xpathText(doc, xpath_abstract, namespace, "\n")``` before running ZU.trimInternal, which will throw an error when run on ```null```.
I'd do 
```
var abstract_note = ZU.xpathText(doc, xpath_abstract, namespace, "\n");
if (abstract_note){
  newItem.abstractNote = ZU.trimInternal(abstract_note);
}
```

File: HCSP.js [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    var pdf_filename_regex = /=(.*)/
+
+    // For each PDF anchor get the url and title and push it
+    for (one_pdf_anchor of xresult_pdfs) {
+        var pdf_url = one_pdf_anchor.href;
+        var pdf_title = pdf_url.match(pdf_filename_regex)[1];
+
+        var one_pdf_attachement = {
+            url: pdf_url,
+            title: pdf_title,
```
Comment (adam3smith) [inline HCSP.js]: As per Zotero conventions, this should be "HCSP Full Text PDF" not the pdf title

File: HCSP.js [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    newItem.publicationTitle = "Rapport de l'HCSP";
+    newItem.language = "fr";
+    newItem.institution = "Haut Conseil de la Santé Publique";
+    newItem.place = "Paris";
+
+    // Creator is the institution (HCSP)
+    // Names of the author are inside the PDFs (if there is one)
+    // TODO: Find a way to have a single field instead of two
+    var creator = ZU.cleanAuthor('HCSP', 'author', false);
+    newItem.creators.push(creator);
```
Comment (adam3smith) [inline HCSP.js]: Use
```newItem.creators.push({"lastName": "HCSP", "creatorType": "author", "fieldMode": 1})```;

File: HCSP.js [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        }
+    }
+    
+    newItem.date = String(date);
+
+    /*---------- Tags ----------*/
+    var xpathTags = '//div[@id="doc-donnees"]/span/a';
+    var xresultTags = ZU.xpath(doc, xpathTags);
+
+    for (var nextResult of xresultTags) {
```
Comment (adam3smith) [inline HCSP.js]: again, use simpler loop.

File: HCSP.js [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+/** BEGIN TEST CASES **/
+var testCases = [
+    {
+        "type": "web",
+        "url": "http://www.hcsp.fr/explore.cgi/avisrapportsdomaine?clefr=509",
+        "items": [
+            {
+                "itemType": "report",
+                "title": "Évaluation des besoins des personnes handicapées et  élaboration des réponses au sein des MDPH",
```
Comment (adam3smith) [inline HCSP.js]: looks like this needs cleaning up using ```ZU.trimInternal()``` -- there's a double space between "et" and "elaboration"

**Comment by adam3smith** [comment]
I think that should cover most issues -- might need another round of reviews after you fix these. Let us know if you have any questions, and, of course, thanks for putting this together.

File: HCSP.js (hunk 2 above)
Comment (aurimasv) [inline HCSP.js]: That is actually a RegExp (strings are converted to RegExp in `match`), but explore.cgi part, and even both "avisrapports..." parts should simply be moved to the target RegExp above.

Also, iirc, we do case insensitive matching with the target RegExp

File: HCSP.js [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    newItem.place = "Paris";
+
+    // Creator is the institution (HCSP)
+    // Names of the author are inside the PDFs (if there is one)
+    // TODO: Find a way to have a single field instead of two
+    var creator = ZU.cleanAuthor('HCSP', 'author', false);
+    newItem.creators.push(creator);
+
+    /*---------- URL ----------*/
+    newItem.url = doc.location.href;
```
Comment (aurimasv) [inline HCSP.js]: Use url passed to `scrape`as the argument

File: HCSP.js [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    for (var nextP of docDonnees) {
+        var pContent = nextP.textContent;
+
+        if (pContent.match("Date du document")){
+            var date = pContent.match("[0-9]{2}/[0-9]{2}/[0-9]{4}");
+            break;
+        }
+    }
+    
+    newItem.date = String(date);
```
Comment (aurimasv) [inline HCSP.js]: (1) while this works (sometimes), please keep the scope of the variables sane. i.e. above, `date`should be declared outside the `for` loop.
(2) because of the for loop and the if statement, `date` could remain `undefined` (and because of (1), undeclared, which iirc throws an error), do you need to check to make sure. Or simply assign `newItem.date` inside the `if`
(3) you're converting an array of strings to string, which probably works out somehow, but you should just retrieve the first (only) element of the array

File: HCSP.js (hunk 1 above)
Comment (zuphilip) [inline HCSP.js]: This means AFAIK
```
^https?://www\.hcsp\.fr/
```

File: HCSP.js [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+                "itemType": "report",
+                "title": "Évaluation des besoins des personnes handicapées et  élaboration des réponses au sein des MDPH",
+                "creators": [
+                    {
+                        "firstName": "",
+                        "lastName": "HCSP",
+                        "creatorType": "author"
+                    }
+                ],
+                "date": "11/05/2015",
```
Comment (aurimasv) [inline HCSP.js]: Ambiguous date. Since you're using a RegExp to extract it anyway, please convert it to `YYYY-MM-DD` format

File: HCSP.js (hunk 1 above)
Comment (jomuller) [inline HCSP.js]: Fixed. Thank you!

File: HCSP.js (hunk 2 above)
Comment (jomuller) [inline HCSP.js]: The `Explore.cgi` is not necessary I think. I will add it to the _target_ of the header. This way, no needs to keep a regex and I can use _.indexOf_ if it's more efficient. What do you think?

Comment (adam3smith) [inline HCSP.js]: yes, perfect

File: HCSP.js (hunk 4 above)
Comment (jomuller) [inline HCSP.js]: Fixed

File: HCSP.js (hunk 10 above)
Comment (jomuller) [inline HCSP.js]: I'm not used with this syntax. Is this a trik to use named argument in javascript? By the way thank you for the advice! Fixed

File: HCSP.js (hunk 8 above)
Comment (jomuller) [inline HCSP.js]: I used the argument `namespace` because I did't find how to use named argument in javascript and I wanted to access to the last argument (the separator). But there must be a more elegant solution.

File: HCSP.js (hunk 10 above)
Comment (adam3smith) [inline HCSP.js]: so ```ZU.cleanAuthor()``` just generates a Zotero creator object -- the same one you can see down in the translator test -- from a string (and some optional arguments like the creator type). But instead, you can just write that object out yourself, which is what I suggested to you here. Among other things, that allows you to just specify the fieldMode rather than adding it later on.

File: HCSP.js (hunk 14 above)
Comment (jomuller) [inline HCSP.js]: Thank you for yours advises. I followed all + turn the date to the ISO 8601 format.

File: HCSP.js [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+    // Names of the author are inside the PDFs (if there is one)
+    newItem.creators.push({"lastName": "HCSP", "creatorType": "author", "fieldMode": 1});
+
+    /*---------- URL ----------*/
+    newItem.url = url;
+    
+    /*---------- Title ----------*/
+    var xpathTitle = '//li[@class="active"]';
+    var title = ZU.xpathText(doc, xpathTitle);
+    newItem.title = title;
```
Comment (adam3smith) [inline HCSP.js]: could you still use ```ZU.trimInternal()``` on the title to prevent double spaces (see below).

File: HCSP.js (hunk 9 above)
Comment (adam3smith) [inline HCSP.js]: this is still not fixed.

File: HCSP.js [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+            mimeType: "application/pdf"
+        }
+
+        newItem.attachments.push(one_pdf_attachement);
+    }
+
+    newItem.complete();
+}
+
+/** BEGIN TEST CASES **/
```
Comment (adam3smith) [inline HCSP.js]: update your test when you're done with everything else.

File: HCSP.js (hunk 9 above)
Comment (jomuller) [inline HCSP.js]: Just working on it :-). There is a little difficulty: there should be 2 PDFs in some cases: the advise and the repport. Then I will search for the text in the `<a>` and then produce "Avis HCSP Full Text PDF" or "Rapport HSCP Full Text PDF". Is it ok?

File: HCSP.js [hunk 18]
```diff
@@ -0,0 +1,231 @@
+{
+    "translatorID": "3e375db7-5f6e-4861-ab8b-3ddb0f4accc9",
+    "label": "HCSP",
+    "creator": "Joris Muller",
+    "target": "^https?//www\.hcsp\.fr/[Ee]xplore.cgi/",
```
Comment (aurimasv) [inline HCSP.js]: Double backslashes if you're escaping inside the string. I'd recommend using Scaffold, which takes care of this. Also, you don't need to worry about case, since we do case-insensitive matching

Comment (aurimasv) [inline HCSP.js]: and the last period is not escaped

File: HCSP.js [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+    "minVersion": "2.1",
```
Comment (aurimasv) [inline HCSP.js]: This should be 3.0

File: HCSP.js (hunk 9 above)
Comment (adam3smith) [inline HCSP.js]: that's perfect yes.

File: HCSP.js [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+    } else if (url.indexOf("avisrapports") != -1) {
+        return 'multiple';
+    }
+}
+
+function doWeb(doc, url) {
+
+    if (detectWeb(doc, url) == "multiple") {
+
+        var allItems = getSearchResults(doc);
```
Comment (aurimasv) [inline HCSP.js]: Since you have it in a function anyway, might as well use it in `detectWeb`. Generally, something like `return getSearchResults(doc) ? 'multiple' : false;`

File: HCSP.js [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+
+        Zotero.selectItems(allItems, function(items) {
+            if (!items) return true;
+
+            var articlesURLs = [];
+            for (var itemURL in items) {
+                articlesURLs.push(itemURL);
+                }
```
Comment (aurimasv) [inline HCSP.js]: nit: indents

File: HCSP.js [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+    
+    /*---------- Title ----------*/
+    var xpathTitle = '//li[@class="active"]';
+    var title = ZU.xpathText(doc, xpathTitle);
+    newItem.title = title;
+ 
+    /*---------- Date ----------*/
+    // Date is not easy to find because it is not in a named bloc. It is
+    // somewhere in one of the <p> in the div 'doc-donnees'
+    // Then to find it, use a Regex on each <p> bloc
```
Comment (aurimasv) [inline HCSP.js]: Since you're not doing anything complicated with your regexp, you can actually incorporate it into your xpath:
```
var date = ZU.xpath(doc, "//div[@id='doc-donnees']/p[contains(text(),'Date du document')][0]");
if (date) {
  date = date.textContent.match(/\b(\d{2})\/(\d{2})\/(\d{4})\b/); // note the \b
  if (date) {
    newItem.date = date[3] + "-" + date[2] + "-" + date[1];
  }
}

File: HCSP.js (hunk 19 above)
Comment (jomuller) [inline HCSP.js]: Thank you for this precision. Fixed.

File: HCSP.js (hunk 18 above)
Comment (jomuller) [inline HCSP.js]: I followed your proposition and used Scaffold. Fixed

File: HCSP.js (hunk 16 above)
Comment (jomuller) [inline HCSP.js]: Fixed

File: HCSP.js (hunk 20 above)
Comment (jomuller) [inline HCSP.js]: Good idea. Added

File: HCSP.js (hunk 22 above)
Comment (jomuller) [inline HCSP.js]: A little bit harder to read for me (then harder to maintain). Does this improve performances?

**Comment by jomuller** [comment]
Thank you very much for your reviews all! I learned a lot. I hope now the translator to be closer to the quality standards of Zotero. I tried to add all your suggestions and added two test with corner cases.

**Comment by adam3smith** [comment]
nothing more from my side, thanks! @aurimasv -- anything from you?

**Comment by aurimasv** [comment]
All looks ok. Could you squash the commits?

**Comment by adam3smith** [comment]
let us know if you need help with rebase&squash. There are instructions e.g. here:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by jomuller** [comment]
I'm a little bit lost in this process. I never did it before, and I'm not sure to understand the goal. Have I to merge all my commits into a single one? Is that right?

**Comment by aurimasv** [comment]
Yes. If you don't want to spend time figuring it out, we can actually do it
during the merge. So up to you

**Comment by jomuller** [comment]
Thank you. I will try to do it. Another occasion to learn something about GIT!

**Comment by jomuller** [comment]
I tried to rebase and squash. First time I did it. Is it what you expected?

**Comment by aurimasv** [comment]
Yeah, that's great. One not though. Could you amend that commit and clean up the commit message? It sound be short and simple (pretty much your first line) http://stackoverflow.com/questions/179123/edit-an-incorrect-commit-message-in-git

**Comment by aurimasv** [comment]
Or... Not

**Comment by adam3smith** [comment]
oops, sorry, too late.

**Comment by jomuller** [comment]
May I fix it on my side or is it really to late ? Whatever, thank you to merge it!

**Comment by aurimasv** [comment]
Yeah, once it's merged into the master branch, it's permanent (for the most part). Thanks for working on this.

**Comment by adam3smith** [comment]
I mean - if it's important you can always create another pull request.

**Comment by jomuller** [comment]
May I fix it on my side by changing the commit message or is it really to late?

**Comment by adam3smith** [comment]
no, not really -- we can technically mess with the commit history in master the same way you did to squash your commits, but I don't think we'd want to do that and you wouldn't be able to do it at all.
